pkg/contexts/ocm/transfer/autohandler.go

```diff
@@ -0,0 +1,111 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Open Component Model contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package transfer
+
+import (
+	"sync"
+
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/transfer/transferhandler"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/transfer/transferhandler/standard"
+	"github.com/open-component-model/ocm/pkg/errors"
+)
+
+func init() {
+	SetDefaultTransferHandlerFactory(func() transferhandler.TransferHandler { h, _ := standard.New(); return h })
+}
+
+var (
+	lock           sync.Mutex
+	defaultHandler func() transferhandler.TransferHandler
+)
+
+func SetDefaultTransferHandlerFactory(f func() transferhandler.TransferHandler) {
+	lock.Lock()
+	defer lock.Unlock()
+	defaultHandler = f
+}
+
+func NewDefaultTransferHandler() transferhandler.TransferHandler {
+	lock.Lock()
+	defer lock.Unlock()
+	return defaultHandler()
+}
+
+// NewTransferHandler creates a transfer handler for the given set of transfer
+// options. If there is no handler type supporting all the given options
+// an ErrNotSupported error is returned.
+// If no handler can be found according to the origin handlers of given options
+// (an option always belongs to a dedicated option set of a dedicated handler
+// but may be supported by other option sets, also) the options sets of all
+// registered handler types with checked by trying the most significant
+// handlers, first, to find the best matching handler for the given option set.
+func NewTransferHandler(list ...transferhandler.TransferOption) (h transferhandler.TransferHandler, ferr error) {
+	var opts transferhandler.TransferOptions
+
+	created := -1
+outer:
+	for {
+		ferr = nil
+		for i := 0; i < len(list); i++ {
+			o := list[i]
+			if o == nil {
+				continue
+			}
+			if opts == nil {
+				created = i
+				opts = o.NewOptions()
+			}
+			if err := o.ApplyTransferOption(opts); err != nil {
+				if errors.IsErrNotSupportedKind(err, transferhandler.KIND_TRANSFEROPTION) {
+					if i <= created {
+						// give a second chance to later options
```

Comment:
Why? What would change between the handlers and the options that would make it pass on a second try? Why not a third or a forth? These seem like an arbitrary thing and doesn't help in understand what's going on here.

Reply:
Depending on the order of the given option list, it might happen that later options relate to a more specialized list.
Therefore every later option, if not matching the actual option target gets a (second compared to the first failing) chance to provide a better option target able to consume this failing option setter.

This means every option has
- a first chance, to configure the already known option target
- and a second chance to provide an own target object, it is not applicable to the actual target

Reply:
This seems like a convoluted way to provide fall-back. But okay.